### ISSUE TITLE
Campaign Contract Specification

### ISSUE THREAD

**Issue by perfectmak** [perfectmak/FDR#1]
# Abstract
The following describes standard functions a Campaign contract should implement.

# Motivation
These will allow dapps and wallets to handle campaigns across multiple interfaces/dapps.

# Specification
More to come after discussions.

**Comment by saopayne** [comment]
### Methods

#### contributeToFundRaise
``` 
function contributeToFundRaise() public  
```
donate to the ongoing fund raise

#### withdrawDonation
```
function withdrawDonation() returns (bool) 
```
get back donation that was made to the campaign

#### changeDeadline
```
 function changeDeadline()
```
either reduce or increase the set deadline for the fund raise

#### changeFundTarget
```
function changeFundTarget(uint amount)
```
change the target amount to meet for the fund raise

#### refundDonation
```
function refundDonation(address recipient, uint amount) public 
```
 refund the funds to the donors if the target is not met or the fund raise is cancelled
or flagged not to be genuine

#### getTotalDonation
```
function getTotalDonation() public 
```
retrieve the total amount donated thus far

#### getDonors
```
function getDonors() public 
```
retrieve the addresses of donors on this fund raise

#### destroy
``` 
function destroy() 
```
release funds in the contract to avoid permanent lockdown

### Events

#### FundRaiseEnded
```
event FundRaiseEnded(address raiser, uint _amount)
```
Triggered when the time set for the fund raise has been reached. If the goal is met, the total amount is transferred to the raiser otherwise, refunds are made to the accounts that made donations

#### FundRaiseTargetMet
```
event FundRaiseTargetMet(address raiser, uint _amount)
```
Triggered when the target has been met so people donating can know and can proceed to donate if they  so desire.

#### Deposit
```    
event Deposit(address _from, uint _amount)
```
Triggered when a deposit has been made to the fund raise

#### Refund
```
event Refund(address _to, uint _amount)
```
Triggered when a refund is to be made if the goal is not met or the donor decides to opt out.

**Comment by saopayne** [comment]
@perfectmak what is missing here?

**Comment by perfectmak** [comment]
Yeah, it looks good... I'm still taking a closer look, but I don't think we need to add private functions to the specification, those will be useful for internal implementation.

**Comment by saopayne** [comment]
Specifications updated. Will take a look at the specs again and raise some possible concerns.

**Comment by perfectmak** [comment]
Alright, I've gone through in more details and here are my comments:

1. First, I think we should make the names less explicit, for example `contributeToFundRaise()` can just be `contribute()`, `withdrawDonation()` can just be `withdraw()`, etc.

2. Also `contributeToFundRaise()` should take a parameter (perhaps uint) specifying the amount of FDR to contribute and therefore should have the `payable` modifier. Also it should return a `bool` cause I assume once the deadline is over, this should either return a `false` or it should just `throw`

3. I don't think `refundDonation(receipient, amount)` required those two parameters, since I assume we would be keeping track of donors and their amount in memory, so we can just use that to refund their FDRS to them.

4. Then, lastly, I think we should not allow `changeFundTarget()` and `changeDeadline()`. I feel those parameters should be passed along with the contract during deployment, so it is fixed. 

Those are my observations, let me know what your take is on them.

**Comment by saopayne** [comment]
1. For the naming, that will be fixed.

2. Absolutely, that makes sense. @perfectmak an update on this: the amount is actually not required, the amount can be gotten via `msg.value` and the address `msg.sender` and the payable modifier has been included.

3. True since we will have the details loaded in the donors list.

4. While I agree with the `changeFundTarget()`, I think `changeDeadline()` should be permissible. An example that comes to mind is if you need to raise funds for a certain event and the event is moved forward, I might want to extend the deadline of the fundraise too. Tentatively, I will remove the two methods.

**Comment by perfectmak** [comment]
Okay, I see your point with the `changeDeadline()`. We can allow that then. 
So great. I think once those are updated, this contract looks okay.
So you can proceed with its dev. I'll also work on getting the FDR tokens contract setup.

**Comment by saopayne** [comment]
Will update the Methods and events shortly, made some changes but doesn't deviate from what we have already. The PR will reflect the changes made.
Looking forward to the tokens contract. Kindly create an issue on that if need be.
